### PR TITLE
HHH-16830 - apply filters to 'find()' method

### PR DESCRIPTION
**Pull request by gregoriopalama**
https://hibernate.atlassian.net/browse/HHH-16830

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/binder/internal/TenantIdBinder.java [hunk 1]
```diff
@@ -54,7 +54,8 @@ public void bind(
 					new FilterDefinition(
 							FILTER_NAME,
 							"",
-							singletonMap( PARAMETER_NAME, tenantIdType )
+							singletonMap( PARAMETER_NAME, tenantIdType ),
+							false
```
[the remark this answers is not in the document]
Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/binder/internal/TenantIdBinder.java:58]: Yeah, you are right. This is the exact case where it should be true as default. I'll update it

[the remark this answers is not in the document]
Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/binder/internal/TenantIdBinder.java]: Done, and updated the tests consequently

**Comment by beikov** [comment]
This will have to wait until https://github.com/hibernate/hibernate-orm/pull/7735 is merged. Also, @sebersole mentioned that he dislikes the annotation member name. Let's wait what he proposes as alternative.

**Comment by hamid646m** [comment]
> This will have to wait until #7735 is merged. Also, @sebersole mentioned that he dislikes the annotation member name. Let's wait what he proposes as alternative.

@beikov i believe that PR has been merge now

**Comment by gregoriopalama** [comment]
@sebersole following your comments on #7735 , show we use the same approach on XSD here? 
What about the annotation member name? 
I can update this accordly.

**Comment by sebersole** [comment]
> What about the annotation member name?

My concern about the annotation attribute name is that I believe this actually affects operations other than just find.  Which is fine as it is really the intent of this request - always apply the filter.  But I think the name is therefore misleading.

I initially thought `#alwaysApply` but I think that it confusing with the concept of auto applied.  

So long story short, I don't have a better suggestion at the moment.  I just don't think `#appliedOnFind` is a good name.

**Comment by sebersole** [comment]
As I don't have a better suggestion right now, one option would be to mark the member as `@Incubating` for 6.5 and move on coming back to the naming later.

Though obviously that's a last resort.

**Comment by sebersole** [comment]
@beikov suggested `#applyToLoadByKey` which I think is reasonable.

**Comment by sebersole** [comment]
Actually after thinking about this some more and discussing within the team, I think we might want something slightly different here.  Though I think the PR as-is is very close and you can easily make it match that.

What we'd prefer to see is :

1. By default apply filters as they were in 5.x, which I think your PR is already doing
2. Allow filters (via `@FilterDef`) to opt in to applying the filter to *top level* find operations.  The implication here being fetching of to-one associations should never be filtered.  I believe this is something we are missing here, right?  "top level" here can/should include natural-id loading, multi id loading and multi natural-id loading.
3. Work on https://hibernate.atlassian.net/browse/HHH-17117 to allow deeper application of "filters".

So given this limited scope, I guess `@FilterDef#applyOnFind` is a good name.  

I'd also like to see explicit tests that ensure fetching to-one associations does not apply the filters.

**Comment by beikov** [comment]
After todays team discussion, we concluded that we prefer the `applyToLoadByKey` member name.
We decided that it is best to always apply filters on finds, regardless of whether it is for an association.
To protect you from accidently destroying your own data (i.e. FKs set to `null`), we decided that it is best to apply the same technique that we use for `@NotFound`, which is to also fetch the FK of an association. This allows us to throw a `NotFoundException` if a filter would filter out the target row for an association.

We need a test that asserts this behavior though i.e. setup data that would filter out the association target row and assert that the exception is thrown.

**Comment by beikov** [comment]
Could you please rename the annotation member and do a rebase @gregoriopalama ?

I'll take care of implementing the part about

> apply the same technique that we use for `@NotFound`

**Comment by gregoriopalama** [comment]
@beikov not sure everything went smooth with the rebase. Before I create issues, may I ask you if I made something wrong? I see complaints from the bot on commits name, but those are from the rebase.

**Comment by beikov** [comment]
The rebase didn't work as you can see on the UI based on the message about conflicting files.

You need to `git fetch` from the "upstream" repository https://github.com/hibernate/hibernate-orm first, then rebase your branch against that upstream main branch. After that, force push to your branch.

**Comment by kevinm416** [comment]
Hi @gregoriopalama, not sure what went wrong with your rebase, but this is what I would do to try to resolve the situation:

```
# Not sure if you need to add a remote, but I did after checking out your repo. 
# If you have an upstream when you run `git remote` then you can skip this.
$ git remote add upstream git@github.com:hibernate/hibernate-orm.git
$ git fetch upstream

# check out new branch based on hibernate's up-to-date main and switch to it
$ git checkout -b main-2 upstream/main
# cherry pick the commits from your branch
$ git cherry-pick 1fd3dcccca14bb890387eafef22034680d366fda
$ git cherry-pick 98d21c937c84db21d9d5660ac2afeb6a208abead
# Resolve conflicts with upstream (there are some small ones), then commit
# Delete your original feature branch
$ git br -D HHH-16830
# Rename the the new branch to match the original branch name
$ git br -m HHH-16830
# Force push your feature branch now that it's up-to-date with upstream/main
$ git push -f HHH-16830
```

I did these steps locally, but obviously can't attempt pushing the results.

**Comment by kingan379** [comment]
Any updates on this MR? It is a big security problem holding many teams back.

**Comment by kevinm416** [comment]
This is a good starting point for how to work around this issue: https://github.com/M-Devloo/Spring-boot-auth0-discriminator-multitenancy/pull/3. I found the CriteriaBuilder approach used in that PR too slow, but it was very helpful to see how to set up a global base repository.

**Comment by karmann-dm** [comment]
Hey folks, any update on the issue?

**Comment by beikov** [comment]
Superseded by https://github.com/hibernate/hibernate-orm/pull/8463

**Pull request closed without merging**